test/integration/apiserver/cel/validatingadmissionpolicy_test.go

```diff
@@ -0,0 +1,942 @@
+	return withPolicyMatch("namespaces", policy)
+}
+
+func withPolicyMatch(resource string, policy *admissionregistrationv1alpha1.ValidatingAdmissionPolicy) *admissionregistrationv1alpha1.ValidatingAdmissionPolicy {
+	policy.Spec.MatchConstraints = &admissionregistrationv1alpha1.MatchResources{
+		ResourceRules: []admissionregistrationv1alpha1.NamedRuleWithOperations{
+			{
+				RuleWithOperations: admissionregistrationv1alpha1.RuleWithOperations{
+					Operations: []admissionregistrationv1.OperationType{
+						"CREATE",
+					},
+					Rule: admissionregistrationv1.Rule{
+						APIGroups: []string{
+							"",
+						},
+						APIVersions: []string{
+							"*",
+						},
+						Resources: []string{
+							resource,
+						},
+					},
+				},
+			},
+		},
+	}
+	return policy
+}
+
+func withPolicyExistsLabels(labels []string, policy *admissionregistrationv1alpha1.ValidatingAdmissionPolicy) *admissionregistrationv1alpha1.ValidatingAdmissionPolicy {
+	if policy.Spec.MatchConstraints == nil {
+		policy.Spec.MatchConstraints = &admissionregistrationv1alpha1.MatchResources{}
+	}
+	matchExprs := buildExistsSelector(labels)
+	policy.Spec.MatchConstraints.ObjectSelector = &metav1.LabelSelector{
+		MatchExpressions: matchExprs,
+	}
+	return policy
+}
+
+func makeBinding(name, policyName, paramName string) *admissionregistrationv1alpha1.ValidatingAdmissionPolicyBinding {
+	return &admissionregistrationv1alpha1.ValidatingAdmissionPolicyBinding{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+		Spec: admissionregistrationv1alpha1.ValidatingAdmissionPolicyBindingSpec{
+			PolicyName: policyName,
+			ParamRef: &admissionregistrationv1alpha1.ParamRef{
+				Name:      paramName,
+				Namespace: "default",
+			},
+		},
+	}
+}
+
+func withBindingExistsLabels(labels []string, policy *admissionregistrationv1alpha1.ValidatingAdmissionPolicy, binding *admissionregistrationv1alpha1.ValidatingAdmissionPolicyBinding) *admissionregistrationv1alpha1.ValidatingAdmissionPolicyBinding {
+	if policy != nil {
+		// shallow copy
+		constraintsCopy := *policy.Spec.MatchConstraints
+		binding.Spec.MatchResources = &constraintsCopy
+	}
+	matchExprs := buildExistsSelector(labels)
+	binding.Spec.MatchResources.ObjectSelector = &metav1.LabelSelector{
+		MatchExpressions: matchExprs,
+	}
+	return binding
+}
+
+func buildExistsSelector(labels []string) []metav1.LabelSelectorRequirement {
+	matchExprs := make([]metav1.LabelSelectorRequirement, len(labels))
+	for i := 0; i < len(labels); i++ {
+		matchExprs[i].Key = labels[i]
+		matchExprs[i].Operator = metav1.LabelSelectorOpExists
+	}
+	return matchExprs
+}
+
+func makeConfigParams(name string, data map[string]string) *v1.ConfigMap {
+	return &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+		Data:       data,
+	}
+}
+
+func checkForFailedRule(t *testing.T, err error) {
+	if !strings.Contains(err.Error(), "failed expression") {
```

Comment:
For a follow-up: given it's possible to know the expected error string, seems like we should just compare the full string